src/loader.ts

```diff
@@ -18,29 +18,92 @@
  * @packageDocumentation
  */
 
+import * as path from 'path';
+import * as semver from 'semver';
+import * as readPkgUp from 'read-pkg-up';
 /**
  * Import function signature type's definition.
  */
 import {HandlerFunction} from './functions';
 
+// Dynamic import function required to load user code packaged as an
+// ES module is only available on Node.js v13.2.0 and up.
+//   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#browser_compatibility
+// Exported for testing.
+export const MIN_NODE_VERSION_ESMODULES = '13.2.0';
+
+/**
+ * Determines format of the given module (CommonJS vs ES module).
+ *
+ * Implements "algorithm" described at:
+ *   https://nodejs.org/api/packages.html#packages_type
+ *
+ * In words:
+ *   1. A module with .mjs extension is an ES module.
+ *   2. A module with .clj extension is CommonJS.
+ *   3. A module with .js extensions where the nearest package.json's
+ *      with "type": "module" is an ES module
+ *   4. Otherwise, it is CommonJS.
+ *
+ * @returns {Promise<'commonjs' | 'module'>} Module format ('commonjs' or 'module')
+ */
+async function moduleFormat(
+  modulePath: string
+): Promise<'commonjs' | 'module'> {
+  if (/\.mjs$/.test(modulePath)) return 'module';
+  if (/\.cjs$/.test(modulePath)) return 'commonjs';
+
+  const pkg = await readPkgUp({
+    cwd: path.dirname(modulePath),
+    normalize: false,
+  });
+
+  // Default to commonjs unless package.json specifies type as 'module'.
+  return pkg?.packageJson.type === 'module' ? 'module' : 'commonjs';
+}
+
+/**
+ * Dynamically load import function to prevent TypeScript from
+ * transpiling into a require.
+ *
+ * See https://github.com/microsoft/TypeScript/issues/43329.
+ */
+const dynamicImport = new Function(
+  'modulePath',
+  'return import(modulePath)'
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+) as (modulePath: string) => Promise<any>;
+
 /**
  * Returns user's function from function file.
  * Returns null if function can't be retrieved.
  * @return User's function or null.
  */
-export function getUserFunction(
+export async function getUserFunction(
   codeLocation: string,
   functionTarget: string
-): HandlerFunction | null {
+): Promise<HandlerFunction | null> {
   try {
     const functionModulePath = getFunctionModulePath(codeLocation);
     if (functionModulePath === null) {
       console.error('Provided code is not a loadable module.');
       return null;
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-var-requires
-    const functionModule = require(functionModulePath);
+    let functionModule;
+    if ((await moduleFormat(functionModulePath)) === 'module') {
```

Comment:
nit: move `await` to it's own statement and assign the result to a local variable